lib/child-writer/index.js

```diff
@@ -165,59 +177,222 @@ exports.write = (image, drive, options) => {
+      image,
+      device: drive.device,
+      validateWriteOnSuccess: options.validateWriteOnSuccess,
+      unmountOnSuccess: options.unmountOnSuccess
     })
 
-    child.stderr.on('data', (data) => {
-      bridgeRobotMessage(data.toString())
+    ipc.serve()
 
-      // This function causes the `close` event to be emitted
-      child.kill()
+    ipc.server.on('error', this.onError)
+    ipc.server.on('message', this.bridgeRobotMessage)
+
+    ipc.server.once('start', () => {
+      this.child = childProcess.fork(CONSTANTS.WRITER_PROXY_SCRIPT, argv, {
+        silent: true,
+        env: process.env
+      })
+
+      this.child.stdout.on('data', (data) => {
+        console.info(`WRITER: ${data.toString()}`)
+      })
+
+      this.child.stderr.on('data', (data) => {
+        this.bridgeRobotMessage(data.toString())
+        this.abort()
+      })
+
+      this.child.on('error', this.onError)
+      this.child.on('exit', this.onExit)
     })
 
-    child.on('error', emitError)
+    ipc.server.start()
 
-    child.on('exit', (code, signal) => {
-      terminateServer()
+    return this
+  }
 
-      if (code === EXIT_CODES.CANCELLED) {
-        return emitter.emit('done', {
-          cancelled: true
-        })
-      }
+  /**
+   * @summary Detach event handlers from the IPC
+   * and child to avoid emitting further events
+   * @private
+   *
+   * @example
+   * this.detachHandlers()
+   */
+  detachHandlers () {
+    // NOTE: The `ipc.server` is not an instance of EventEmitter,
+    // but a PubSub; see https://github.com/RIAEvangelist/event-pubsub
+    // Thus the method to remove event handlers is not `.removeListener()`
+    ipc.server.off('error', this.onError)
+    ipc.server.off('message', this.bridgeRobotMessage)
+    this.child.removeListener('exit', this.onExit)
+  }
+
+  /**
+   * @summary Abort the current writing process
+   * @public
+   *
+   * @param {Function} callback - callback
+   *
+   * @example
+   * childWriter.abort(() => {
+   *   console.log('Writing aborted')
+   * })
+   */
+  abort (callback) {
+    debug('abort')
 
-      // We shouldn't emit the `done` event manually here
-      // since the writer process will take care of it.
-      if (code === EXIT_CODES.SUCCESS || code === EXIT_CODES.VALIDATION_ERROR) {
-        return null
+    if (this.aborting) {
+      debug('abort:in-progress')
+      if (_.isFunction(callback)) {
+        this.once('close', callback)
       }
+      return
+    }
 
-      const error = new Error(`Child process exited with code ${code}, signal ${signal}`)
-      error.code = code
-      error.signal = signal
+    this.aborting = true
+    this.detachHandlers()
+    this.child.once('exit', (code, signal) => {
+      debug('abort:child:exit', code, signal)
+      this.emit('done', {
+        cancelled: true
+      })
+      this.terminateServer(callback)
+    })
 
-      return emitError(error)
+    ipc.server.broadcast('abort', {
+      signal: 'SIGINT'
```

Comment:
Can we use the constants from `os.constants.signals`?

Reply:
subprocess.kill([signal]) expects signal to be a string, not a number (which os.constants.signals.SIGINT is), so that's not going to work I'm afraid